File: Source/C#/RiceDoctor/RiceDoctor.OntologyManager/Models/Attribute.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;
using RiceDoctor.Shared;
using static RiceDoctor.OntologyManager.GetType;

namespace RiceDoctor.OntologyManager
{
    public class Attribute : Entity<Attribute>
    {
        [CanBeNull] private IReadOnlyCollection<Class> _allDomains;
        private bool _canGetAllDomains;
        private bool _canGetComment;
        private bool _canGetDirectDomains;
        [CanBeNull] private string _comment;
        [CanBeNull] private IReadOnlyCollection<Class> _directDomains;

        public Attribute(
            [NotNull] string id,
            [CanBeNull] string label = null,
            [CanBeNull] string comment = null,
            [CanBeNull] DataType? range = null,
            [CanBeNull] IReadOnlyList<string> enumeratedValues = null)
            : base(id, label)
        {
            if (comment != null)
            {
                _comment = comment;
                _canGetComment = false;
            }

            Range = range;
            if (Range == DataType.Enumerated)
            {
                Check.NotEmpty(enumeratedValues, nameof(enumeratedValues));
                EnumeratedValues = enumeratedValues;
            }
        }

        public DataType? Range { get; }

        [CanBeNull]
        public IReadOnlyList<string> EnumeratedValues { get; }

        public override EntityType Type => EntityType.Attribute;

        [CanBeNull]
        public string GetComment()
        {
            if (_canGetComment) return _comment;

            _comment = Manager.Instance.GetComment(Id);
            _canGetComment = true;

            return _comment;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetDirectDomains()
        {
            if (_canGetDirectDomains) return _directDomains;

            _directDomains = Manager.Instance.GetAttributeDomains(Id, GetDirect);
            _canGetDirectDomains = true;

            return _directDomains;
        }

        public void SetDirectDomains([CanBeNull] IReadOnlyCollection<Class> directDomains)
        {
            if (_canGetDirectDomains)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetDirectDomains)));

            _directDomains = directDomains;
            _canGetDirectDomains = true;
        }

        [CanBeNull]
        public IReadOnlyCollection<Class> GetAllDomains()
        {
            if (_canGetAllDomains) return _allDomains;

            _allDomains = Manager.Instance.GetAttributeDomains(Id, GetAll);
            _canGetAllDomains = true;

            return _allDomains;
        }

        public void SetAllDomains([CanBeNull] IReadOnlyCollection<Class> allDomains)
        {
            if (_canGetAllDomains)
                throw new InvalidOperationException(CoreStrings.CannotSetAgain(nameof(SetAllDomains)));

            _allDomains = allDomains;
            _canGetAllDomains = true;
        }
    }
}

File: Source/Java/src/main/java/vn/edu/uit/ontologymanager/model/Attribute.java
package vn.edu.uit.ontologymanager.model;

import com.google.gson.annotations.SerializedName;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import vn.edu.uit.shared.Buildable;

import java.util.List;
import java.util.Set;

import static vn.edu.uit.ontologymanager.model.DataType.ENUMERATED;
import static vn.edu.uit.ontologymanager.model.EntityType.ATTRIBUTE;

public class Attribute extends Entity {

    @Nullable
    @SerializedName("DirectDomains")
    private final Set<Class> directDomains;

    @Nullable
    @SerializedName("AllDomains")
    private final Set<Class> allDomains;

    @Nullable
    @SerializedName("Range")
    private final DataType range;

    @Nullable
    @SerializedName("EnumeratedValues")
    private final List<String> enumeratedValues;

    Attribute(@NotNull final Builder builder) {
        super(ATTRIBUTE, builder.id, builder.label);
        directDomains = builder.directDomains;
        allDomains = builder.allDomains;
        range = builder.range;
        enumeratedValues = builder.enumeratedValues;
    }

    public static class Builder implements Buildable<Attribute> {

        @NotNull
        private final String id;

        @Nullable
        private final String label;

        @Nullable
        private Set<Class> directDomains;

        @Nullable
        private Set<Class> allDomains;

        @Nullable
        private DataType range;

        @Nullable
        private List<String> enumeratedValues;

        public Builder(@NotNull final String id, @Nullable final String label) {
            this.id = id;
            this.label = label;
        }

        @NotNull
        public Builder directDomains(@Nullable final Set<Class> directDomains) {
            this.directDomains = directDomains;
            return this;
        }

        @NotNull
        public Builder allDomains(@Nullable final Set<Class> allDomains) {
            this.allDomains = allDomains;
            return this;
        }

        @NotNull
        public Builder range(@Nullable final DataType range) {
            if (range != ENUMERATED) this.range = range;
            return this;
        }

        @NotNull
        public Builder enumeratedValues(@NotNull final List<String> enumeratedValues) {
            this.range = ENUMERATED;
            this.enumeratedValues = enumeratedValues;
            return this;
        }

        @Override
        @NotNull
        public Attribute build() {
            return new Attribute(this);
        }
    }
}
